### ISSUE TITLE
Proposal: let setState accept a function

### ISSUE THREAD

**Issue by cfenzo** [developit/unistore#17]
Like in (p)react, it would be nice if `setState` could accept a function, to prevent state race conditions in async operations...

See https://codesandbox.io/s/qlqr4xl24j for proposed solution..
PR can be provided..

**Comment by developit** [comment]
Would be curious about the filesize hit.  How does that help prevent race conditions for async?

**Comment by cfenzo** [comment]
It's one additional line in `setState()`:
 `if (typeof update === "function") update = update(state);`
It hopefully won't add too much to the file size.

Btw, I managed to contaminate the linked codesandbox with some render-props tinkering, I've reset it to only include the setState change, will get back to the render props later..

Race conditions might've been the wrong term.. But having an async operation set the new state based on (possibly) stale state, causes "race conditions".

**Comment by developit** [comment]
Right, I think I see where you're saying it could help.  Personally I think that's solved via the store already being available though:

```js
const actions = store => ({
  asyncAction(state) {
    setTimeout( () => {
      // using getState here makes the operation atomic:
      let { count } = store.getState();
      store.setState({ count: count + 1 });
    });
  }
})
```

**Comment by cfenzo** [comment]
Off course, that solves the issue quite nicely! It's so obvious... It was right there all the time...

Regarding async/promise, there are probably more scenarios that I can't think of, but as long as you don't `return await`, you could always compare with the current state before returning anything.. Or even use `return await` if you pass the store to the async function.. All though the latter tightly couples the async function with unistore...

I don't see the need for spending extra bytes on a function invocation when it can be solved that easily, unless it later turns out to be a so frequently used pattern that it saves more overall bytes to add it.

**Comment by developit** [comment]
Glad I could help! haha.  Also I meant to mention before - "actions" as defined in unistore already work as setState callbacks in Preact:

```js
const actions = {
  increment(state) {
    return { count: state.count + 1 };
  }
};

const Global = connect('count', actions)( ({ count, increment }) =>
  <button onClick={increment}>Clicks: {count}</button>
);

class Local extends Component {
  increment = this.setState.bind(this, actions.increment);
  render(props, { count }) {
    return <button onClick={this.increment}>Clicks: {count}</button>
  }
}
```

**Comment by grohlingr** [comment]
I would like to reopen this issue. It is convenient to pass update fn to `setState` as is supported in [React hooks setState API](https://reactjs.org/docs/hooks-reference.html#functional-updates).

The change is simple and allows me to write more declarative code.

In my current project, I have state updater functions, which are curried function accepting values to be set and `state` as last argument. These functions maintains shape of the state for me so I don't need to remember paths.

Compare usage with current API

```javascript
import { appendPhotos } from "/utils/store.js"

connect(null, store => ({
  addPhoto: async (_, newFiles) => {
    // some async actions

    // store.getState() must be called to obtain current state
    const state = store.getState()
    const newState = appendPhotos(newFiles, state)
    store.setState(newState)
  }
}))
```

if `setState` would accept state updater fn I could simplify this code to

```javascript
import { appendPhotos } from "/utils/store.js"

connect(null, store => ({
  addPhoto: async (_, newFiles) => {
    // some async actions
    store.setState(appendPhotos(newFiles))
  }
}))
```

here `appendPhotos` is automatically curried so calling it with single argument produces function with signature `state -> newState` i.e. state updater. I personally think it is easier to read. If I want to update more items in store, I simply import state updater functions and call them in one go

```javascript
import { appendPhotos, storageLoading } from "/utils/store.js"

...
...
store.setState(pipe(appendPhotos(newFiles), storageLoading(false)))
```
I enhanced unistore store in my project so it supports state updater functions, but at first, I wrote logic without even looking in the docs, blindly expecting that unistore supports such feature. I was surprised it does not, but given its minimalistic philosophy, I understand. Unistore is well written and easy to extend and I like it. But I think that unification with React hooks API will help many people feel familiar.



PS: [here is my project](https://github.com/grohlingr/travel-and-tell/blob/master/wwwroot/utils/store.js) to illustrate my intended usage
